src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,146 @@
+-- -- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+-- This module can be used to parse a path based on a yang schema (or it's
+-- derivative grammar) and produce a lua table which is a native lua way
+-- of representing a path. The path provided is a subset of XPath supporting
+-- named keys such as [addr=1.2.3.4] and also basic positional querying
+-- for arrays e.g [position()=1] for the first element.
+--
+-- The structure of the path is dependent on the type the node is, the
+-- conversions are as follows:
+--
+-- Scala fields:
+--    A lua string of the member name
+-- Struct fields:
+--    A lua string of the member name
+-- Array fields:
+--    This is a table which has a "name" property specifying member
+--    name and a "key" field which is a 1 based integer to specify the
+--    position in the array.
+-- Table fields:
+--    This is a table which has a "name" property specifying member
+--    name and has a "keys" (not key) property which is either:
+--       - A string representing the key if the table is string keyed.
+--       - A lua table with corrisponding leaf names as the key and the
+--         value as the value.
+module(..., package.seeall)
+
+local schema = require("lib.yang.schema")
+local data = require("lib.yang.data")
+
+local function extract_parts(fragment)
+   local rtn = {query={}}
+   rtn.name = string.match(fragment, "([^%[]+)")
+   for k,v in string.gmatch(fragment, "%[([^=]+)=([^%]]+)%]") do
+      rtn.query[k] = v
+   end
+   return rtn
+end
+
+local handlers = {}
+function handlers.scalar(fragment, tree)
+   return fragment.name, tree
+end
+function handlers.struct(fragment, tree)
+   return fragment.name, tree.members
+end
+function handlers.table(fragment, tree)
+   if #tree.keys == 1 then
+      -- Check if the key is a string based key
```

Comment:
this will fail for tables that have 1 key but are not string-keyed.  the test for tree.string_key is sufficient